crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -860,21 +860,39 @@ pub struct EntityWorldMut<'w> {
 }
 
 impl<'w> EntityWorldMut<'w> {
+    #[track_caller]
+    #[inline(never)]
+    #[cold]
+    fn panic_despawned(&self) -> ! {
+        panic!("Entity {} has been despawned, possibly by hooks or observers, so must not be accessed through EntityWorldMut after despawn.", self.entity);
+    }
+
+    #[inline(always)]
+    #[track_caller]
+    fn assert_not_despawned(&self) {
+        if self.location.archetype_id == ArchetypeId::INVALID {
```

Comment:
This could use the `is_despawned()` method if you wanted.  

```suggestion
        if self.is_despawned() {
```